### ISSUE TITLE
More info about 'different functions with the same name' 

### ISSUE THREAD

**Issue by joshuaiz** [welldone-software/why-did-you-render#33]
Hello,

Love this plugin but the information in the readme and in the blog post about 'different functions with the same name' is really sparse.

Can you explain what this means a bit more in the docs and how to fix it? 

I'm getting these warnings a lot and since the functions are the same (not different functions) it's hard to figure out what to do about it.

A little more information about this particular warning would be super helpful.

Thanks!

**Comment by vzaidman** [comment]
It means that you do something like:
`<PureComponent onClick={function(){console.log('hi')}}/>`

It doesn't makes sense since `PureComponent` will re-render every single time it father re-renders as opposed to:
const fn = function(){console.log('hi')}
`<PureComponent onClick={fn}/>`

that will render only once and won't re-render since onClick is always the same prop.

**Comment by joshuaiz** [comment]
I saw that in the blog post but that doesn't cover my situation...I don't have any functions like that. 

This is using hooks with just a regular `onClick={myfunction}` and not `onClick={() => myfunction()}`

Yet I'm still getting this warning a lot.

**Comment by vzaidman** [comment]
in what stage "myfunction" is created?
please provide more information.

**Comment by joshuaiz** [comment]
` myfunction` would be in the top level `App` component and I'm passing the function down through props to child components.

But these functions are not even being triggered and getting the 'different functions with the same name` warning.

There are *not* different functions with the same name...there is only one function with a particular name so not sure why this warning is being given or what to do about it. 

All that said, the component *is* being re-rendered even with `React.memo` so there is a bottleneck somewhere — I'm just trying to locate it.

**Comment by joshuaiz** [comment]
Here is a screenshot. The `previous` and `next` function *are the same function* so why is it saying 'different functions with the same name? Both functions are at App.js:134 so they are the same function.

![Screen Shot 2019-05-16 at 11 03 48 PM](https://user-images.githubusercontent.com/1096738/57902502-05f13280-782f-11e9-8c72-8dac14efdcbc.png)

**Comment by vzaidman** [comment]
can you add a code of the creation of `myfunction` in `App`?

**Comment by joshuaiz** [comment]
This is the `handleFavorites` function:

```javascript
let newFavorites = []
let found

const handleFavorites = color => {
    if (!color) {
        return
    }
    if (favorites && favorites.length) {
        // check if color is already a favorite
        found = favorites.some(el => el.hex === color.hex)
    }
    if (!found && !favorites.length) {
        newFavorites = [color, ...favorites]
        setFavorites(newFavorites)
        localStorage.setItem('hexy_favorites', JSON.stringify(newFavorites))
    } else if (!found && favorites.length < 15) {
        newFavorites = [color, ...favorites]
        setFavorites(newFavorites)
        localStorage.setItem('hexy_favorites', JSON.stringify(newFavorites))
    } else if (favorites && favorites.length === 15) {
        alert('The maximum number of favorites is 15.')
    }
}
```

And this is the `removeFavorite` function:

```javascript
const removeFavorite = color => {
    if (!color || !favorites) {
        return
    }
    let filteredFavorites = favorites.filter(item => item.hex !== color.hex)
    if (!filteredFavorites.length) {
        clearFavorites()
        setFavoriteSwatches([])
    } else {
        setFavorites(filteredFavorites)
        localStorage.setItem(
            'hexy_favorites',
            JSON.stringify(filteredFavorites)
        )
    }
}
```

Then later in my `App` component, I pass these functions (as well as other props) down to a `<Colors />` component:

```javascript
<Route
    exact
    path="/colors"
    render={() => (
        <Colors
            colors={colors}
            searchInput={searchInput}
            searchSubmitted={searchSubmitted}
            noMatch={noMatch}
            handleFavorites={handleFavorites}
            removeFavorite={removeFavorite}
            favorites={favorites}
            favoriteSwatches={favoriteSwatches}
            setFavoriteSwatches={setFavoriteSwatches}
            handleBright={handleBright}
            sortBright={sortBright}
            getRandoms={getRandoms}
        />
    )}
/>
```

It's the `handleFavorites` and `removeFavorites` functions that are causing the warnings.

But I think I figured it out...I have to try something...I'll report back.

**Comment by joshuaiz** [comment]
So no joy.

In the `<Colors />` component, those functions get passed down to a `<SwatchList />` component which in turn pass them down to a `<Swatch />` component where they can be triggered.

In `Colors`:
```javascript
<SwatchList
    noMatch={noMatch}
    colors={colors}
    handleFavorites={handleFavorites}
    removeFavorite={removeFavorite}
    favorites={favorites}
    favoriteSwatches={favoriteSwatches}
    setFavoriteSwatches={setFavoriteSwatches}
    searchSubmitted={searchSubmitted}
/>
```

In `SwatchList`:
```javascript
<Swatch
    key={color.hex + index}
    color={color}
    index={index}
    handleFavorites={handleFavorites}
    removeFavorite={removeFavorite}
    favorites={favorites}
    isFavorite={isFavorite ? true : false}
/>
```

And finally these are triggered by click events in the `Swatch` component.

But it is only the `SwatchList` component that is re-rendering which is not even using the `handleFavorites` or `removeFavorite` functions at all.

Why?

**Comment by vzaidman** [comment]
in what stage is "handleFavorites" created?

is the code
```
const handleFavorites = color => {
    if (!color) {
        return
    }
    if (favorites && favorites.length) {
        // check if color is already a favorite
        found = favorites.some(el => el.hex === color.hex)
    }
    if (!found && !favorites.length) {
        newFavorites = [color, ...favorites]
        setFavorites(newFavorites)
        localStorage.setItem('hexy_favorites', JSON.stringify(newFavorites))
    } else if (!found && favorites.length < 15) {
        newFavorites = [color, ...favorites]
        setFavorites(newFavorites)
        localStorage.setItem('hexy_favorites', JSON.stringify(newFavorites))
    } else if (favorites && favorites.length === 15) {
        alert('The maximum number of favorites is 15.')
    }
}
```
is just in the global scope of the application? or in `<App/>` or something?

**Comment by joshuaiz** [comment]
Both `handleFavorites` and `removeFavorite` are in `<App />`

**Comment by vzaidman** [comment]
so when `App` re-renders, it re-creates these functions and passes down these function as **new functions with the same name**

please provide the App code that generates them.

I think the [`useCallback` hook](https://reactjs.org/docs/hooks-reference.html#usecallback) might help you.

**Comment by joshuaiz** [comment]
Not sure what you mean by the 'App code that generates them`. They are just functions in my `<App />` component which I provided above.

What I'm not following is that I have `whyDidYouRender` on the `App` component as well and it is not re-rendering. I'm confused as to why these functions are being recreated. Only the `<SwatchList />` component is being re-rendered.

All that said, I think I need to do something with `Context` and `useCallback` as explained here:
https://medium.com/trabe/passing-callbacks-down-with-react-hooks-4723c4652aff

Basically my app is pretty responsive on most pages but any page with a `<SwatchList />` component makes input into the search bar really sluggish because of these re-renders so trying to figure this out.

**Comment by joshuaiz** [comment]
The code is here if you want to look:

https://github.com/joshuaiz/hexy2

**Comment by vzaidman** [comment]
ok, so every time you `setColors` for example, or trigger another setState hook, your `App` is re-rendered.

It doesn't trigger a `whyDidYouRender` warning because it's a "justified re-render" because something actually changed. The App component should re-render and it's fine. This is how React works.

Upon this justified `App` re-render, not only your functions like `getRandoms` and `handleFavorites` are re-created, also `newFavorites` is set to be `[]` again and `found` is reset to be undefined.

I suggest the following:
1. brake App into smaller pieces.
1. put `newFavorites` and `found` into "setState" hooks
1. wrap your functions in `useCallback` as following:
```
 // handle returning search results and updating color list
const handleSearch = useCallback(
  event => {
      event.preventDefault()
      ...
  },
  [a, b], // list of params on which the callback should be recreated, this array might also stay blank
);
```

**Comment by joshuaiz** [comment]
Got it. Thanks for taking a look. That makes a lot of sense so will give that a go.

**Comment by joshuaiz** [comment]
Just FYI, putting `handleFavorites` and `removeFavorite` in `useCallback` fixed it! Thanks.

**Comment by vzaidman** [comment]
I'm really glad to see how this tool helps people to understand how React works :)
spread the love by starring and sharing this project with friends and you are always welcome to contribute to the project as well :)

**Comment by joshuaiz** [comment]
I have and I will - it really helps a lot.

**Comment by Yugank16** [comment]
Can you give me any direction with this, I am getting same warning for handleClick function for the following code in my component:

{members.map((data) => (
            <div key={data.id}>
              <PatientCard
                id={data.id}
                name={`${data.firstName} ${data.lastName}`}
                handleClick={() => handleClick(data)}
              />
            </div>
          ))}


Thanks

**Comment by vzaidman** [comment]
handleClick is indeed a new `() => handleClick(data)` function on each render.

why do you track `PatientCard`? it sounds like there's no problem for it to re-render again.

**Comment by Yugank16** [comment]
I am tracking re-renders for every component and trying to avoid them. Isn't unnecessary re-rendering always a problem?

**Comment by vzaidman** [comment]
No. Only in relatively big components. It's not advised to track everything because it will result in many unneeded "optimizations" that not necessary help.

Better use it to track only pure components and only make key components pure:
```
if (process.env.NODE_ENV === 'development') {
  const whyDidYouRender = require('@welldone-software/why-did-you-render');
  whyDidYouRender(React, {
    trackAllPureComponents: true,
  });
}
```

**Comment by deefisher** [comment]
I'm also getting `different functions with the same name`, however I am using the useCallback hook. Any tips on how to optimise this would be greatly appreciated:

![image](https://user-images.githubusercontent.com/31048185/96585081-ffde6b80-12d6-11eb-951f-84a0c168f58a.png)

**Main component:**
```
const BrokerSearchPageContents = React.memo(({ errorMessage }) => {

    const size = React.useContext(ResponsiveContext);

    const {
        onSearchBarChanged,
        onFilterTagClick,
        onDataSetTagClick,
        onCheckBoxClicked,
        chosenTags,
        filterMenuItems,
        checkedMenuFilters,
        dataSets,
        searchBarInput
    } = useBrokerSearchPageContents();

    const isFiltersEmptyAndNoDataYet = () => {
        return (((checkedMenuFilters && checkedMenuFilters.length === 0) && (searchBarInput && searchBarInput === '')) &&
            ((dataSets && dataSets.length === 0) || (filterMenuItems && filterMenuItems.length === 0))) ? true : false
    }

    const styleRef = React.useRef({ width: '330px', height: '100%' })

    return (
        isFiltersEmptyAndNoDataYet() ?
            <div></div>
            :
            <GenericPageContentHeader
                title='Search datasets'
                subTitle='Search all datasets.'
                customComponent={
                    <BrokerSearchComponent
                        listStateArray={dataSets}
                        onSearchBarChanged={onSearchBarChanged}
                    />
                }
            >
                <Box
                    direction='row'
                    gap='medium'
                    fill
                >
                    {size !== 'small' &&
                        <StyledScrollbar
                            style={styleRef.current}
                        >
                            <Box pad={{ right: '2px' }} gap={!(chosenTags && chosenTags.length > 0) ? '0px' : 'small'}>
                                <TagBox
                                    isHidden={!(chosenTags && chosenTags.length > 0) ? true : false}
                                    chosenTags={chosenTags}
                                    onFilterTagClick={onFilterTagClick}>

                                </TagBox>
                                <SearchMenuComponent
                                    items={filterMenuItems}
                                    onCheckBoxClicked={onCheckBoxClicked}
                                    checkedMenuFilters={checkedMenuFilters}
                                />
                            </Box>
                        </StyledScrollbar>
                    }
                    <BrokerSearchPageList
                        dataSets={dataSets}
                        errorMessage={errorMessage}
                        onTagClick={onDataSetTagClick}
                    />
                </Box>
            </GenericPageContentHeader >
    )
})

export default BrokerSearchPageContents;


BrokerSearchPageContents.propTypes = {
    errorMessage: PropTypes.string,
};

BrokerSearchPageContents.whyDidYouRender = true;
```

**The useBrokerSearchPageContents (with useCallbacks) being used by the main component:**

```
const useBrokerSearchPageContents = () => {

    const [chosenTags, setChosenTags] = React.useState([]);
    const [dataSets, setDataSets] = React.useState(store.getState().dataSets);
    const [searchBarInput, setSearchBarInput] = React.useState('');
    const [filteredDataSets, setFilteredDataSets] = React.useState(store.getState().dataSets);
    const [filterMenuItems, setFilterMenuItems] = React.useState(store.getState().filterMenuItems);
    const [checkedMenuFilters, setCheckedMenuFilters] = React.useState([]);


    const isAllFiltersEmpty = React.useCallback(
        () => {
            if (searchBarInput === '' && checkedMenuFilters && checkedMenuFilters.length === 0) {
                return true;
            }
            else {
                return false;
            }
        },
        [searchBarInput, checkedMenuFilters],
    )


    const handleDataFiltering = React.useCallback(
        () => {
            return filteredDataSets.reduce((acc, cv) => {
                let result = false;

                arrayToLowerCase(checkedMenuFilters).forEach(filterName => {
                    if (cv.title.toLowerCase().includes(filterName.toLowerCase())) {
                        result = true;
                    }
                });
                if (searchBarInput !== '' && cv.title.toLowerCase().includes(searchBarInput)) {
                    result = true;
                }

                return result ? [...acc, cv] : acc;
            }, []);
        },
        [checkedMenuFilters, searchBarInput, filteredDataSets],
    )


    const handleFilterChange = React.useCallback(
        () => {
            const allData = store.getState().dataSets;
            filteredDataSets && filteredDataSets.length === 0 && setFilteredDataSets(allData);

            let newDataSets;
            if (isAllFiltersEmpty()) {
                newDataSets = allData;
            }
            else {
                newDataSets = handleDataFiltering();
            }

            !doArraysMatch(dataSets, newDataSets) && setDataSets(newDataSets);
        },
        [dataSets, handleDataFiltering, isAllFiltersEmpty, filteredDataSets],
    );


    const onFilterTagClick = React.useCallback(
        (tagName) => () => {
            setChosenTags(chosenTags.filter(chosenTagName => chosenTagName !== tagName));
            setCheckedMenuFilters(checkedMenuFilters.filter(checkedMenuFilter => checkedMenuFilter !== tagName));

            const filteredData = returnMissingTagsToDataSet(tagName, filteredDataSets);
            !doArraysMatch(filteredDataSets, filteredData) && setFilteredDataSets(filteredData);
            !doArraysMatch(dataSets, filteredData) && setDataSets(filteredData);
            handleFilterChange(tagName);
        },
        [chosenTags, checkedMenuFilters, filteredDataSets, dataSets, handleFilterChange],
    );


    const onDataSetTagClick = React.useCallback(
        (tagName) => () => {

            let filteredData = disableClickedFilterItemFromDataSet(tagName, filteredDataSets);
            !doArraysMatch(filteredDataSets, filteredData) && setFilteredDataSets(filteredData);
            !doArraysMatch(dataSets, filteredData) && setDataSets(filteredData);

            const lowerCaseCheckedMenuFilters = arrayToLowerCase(checkedMenuFilters);

            setCheckedMenuFilters(checkedMenuFilters && !lowerCaseCheckedMenuFilters.includes(tagName) && [...lowerCaseCheckedMenuFilters, tagName])

            !chosenTags.includes(tagName) && setChosenTags([...chosenTags, tagName]);
            handleFilterChange(tagName);
        },
        [chosenTags, filteredDataSets, checkedMenuFilters, dataSets, handleFilterChange],
    );


    React.useEffect(() => {
        handleFilterChange()
    }, [checkedMenuFilters, searchBarInput, handleFilterChange])


    const onCheckBoxClicked = React.useCallback(
        ({ value, option }) => {
            setCheckedMenuFilters(value);

            setChosenTagsOnCheckBoxClicked({ chosenTags, setChosenTags }, option);

            let filteredData = handleDataSetTagsOnCheckBoxClick({ value, option }, filteredDataSets);
            !doArraysMatch(filteredDataSets, filteredData) && setFilteredDataSets(filteredData);
            !doArraysMatch(dataSets, filteredData) && setDataSets(filteredData);
        },
        [chosenTags, filteredDataSets, dataSets],
    );


    const onSearchBarChanged = React.useCallback(
        (e) => {
            const input = e.target.value.toLowerCase();
            setSearchBarInput(input);

            if (isAllFiltersEmpty()) {
                const allData = store.getState().dataSets;
                !doArraysMatch(dataSets, allData) && setDataSets(allData);
            }
            else {
                handleFilterChange(input);
            }
        },
        [handleFilterChange, isAllFiltersEmpty, dataSets],
    )


    const handleChange = React.useCallback(
        () => {
            const allDataSets = store.getState().dataSets;
            const allFilterMenuItems = store.getState().filterMenuItems.map(
                filterMenuItem => ({ ...filterMenuItem, CheckBoxGroupArray: arrayToLowerCase(filterMenuItem.CheckBoxGroupArray) })
            );
            !doArraysMatch(dataSets, allDataSets) && setDataSets(allDataSets);
            !doArraysMatch(filterMenuItems, allFilterMenuItems) && setFilterMenuItems(allFilterMenuItems);
        },
        [dataSets, filterMenuItems],
    );


    React.useEffect(() => {
        const unsubscribe = store.subscribe(handleChange)
        return () => {
            unsubscribe()
        }
    }, [handleChange])

    
    return {
        onSearchBarChanged,
        onFilterTagClick,
        onDataSetTagClick,
        onCheckBoxClicked,
        chosenTags,
        filterMenuItems,
        checkedMenuFilters,
        dataSets,
        searchBarInput
    }
}

export default useBrokerSearchPageContents
```

**Comment by vzaidman** [comment]
try version `v5.0.0-rc.1`

**Comment by tolgaatrgt** [comment]
I also have a really big rerender problem in every cell of the table because of the handleClick function.

My cell component like 

```
    <StyledTableCell onContextMenu={() => handleClick} style={{ backgroundColor: `${(flash && data) ? 'yellow' : color}`, transition: `${flash ? 'none' : '1s'}` }}>
      <div className='cell-size' >
        {data}
      </div>
    </StyledTableCell >
```

and this cell goes to the upper level like ;

`<MemoizedDataCell handleClick={() => onCellClick(selectedOrders, i, Boolean(val))} data={val} key={colName + i} color={cellColor(i, z)} />`

And finally, in the top parent, i use this function for child ;

```
  const onCellClick = React.useCallback((myOrder: any, i: number, isFilled: boolean) => {

    setCanDeal(isFilled)
    let side = i < 3 ? 'bids' : 'asks'
    console.log(myOrder)
    if (i < 0) {
      return null
    }
    if (!myOrder.asks && !myOrder.bids) {
      const detail = {
        instrName: myOrder.instName, maturity: myOrder.maturity, side: side.charAt(0).toUpperCase() + side.slice(1, -1), contract: `${myOrder.instName} ${myOrder.maturity}`,
        price: '0', quantity: '0', totalQuantity: '0', tradableQuantity: '0'
      }
      return setOrderDetail(detail)
    }

    let orderID = myOrder[`${side}`][0]?.OrderID
    let instrName = myOrder[`${side}`][0]?.InstName
    let maturity = myOrder[`${side}`][0]?.FirstSequenceItemName
    let contract = `${instrName} ${maturity}`
    let broker = myOrder[`${side}`][0]?.Broker
    let price = myOrder[`${side}`][0]?.Price
    let quantity = myOrder[`${side}`][0]?.Volume
    let timeInForce = myOrder[`${side}`][0]?.OrderType.replace(/([A-Z])/g, ' $1').trim()
    let lastUpdTime = new Date(myOrder[`${side}`][0]?.DateTime).toLocaleString()
    let tradableQ = '-'
    let totalQ = '-'
    let priceStat = myOrder[`${side}`][0]?.Status
    side = side.charAt(0).toUpperCase() + side.slice(1, -1)

    const detail = {
      orderID: orderID, instrName: instrName, maturity: maturity, side: side, contract: contract,
      price: price, priceStatus: priceStat,
      quantity: quantity, totalQuantity: totalQ,
      tradableQuantity: tradableQ, broker: broker, lastUpdate: lastUpdTime, timeInForce: timeInForce
    }
    setOrderDetail(detail)
  }, [])

```

Any help ll be appreciated. Thank you, Vitali.
